pingcap/pd
# pd-ctl: add auto-completion for pd-ctl

UCP: #2170 

### What problem does this PR solve? 

fixed: #2170

### What is changed and how it works?

Added new subcommand `completion` 

```
Usage:
    pd-ctl completion SHELL   
```

### Check List 

Tests

 - Manual test (add detailed scripts or steps below)
 - No code

 1. Check shell version and install bash-completion

```shell
Qiannan-MBP:bin qiannan$ echo ${BASH_VERSION}
3.2.57(1)-release
Qiannan-MBP:bin qiannan$ brew info bash-completion
bash-completion: stable 1.3 (bottled)
Programmable completion for Bash 3.2
https://salsa.debian.org/debian/bash-completion
Conflicts with:
  bash-completion@2 (because Differing version of same formula)
/usr/local/Cellar/bash-completion/1.3_3 (189 files, 607.9KB) *
  Poured from bottle on 2020-03-29 at 16:27:16
From: https://github.com/Homebrew/homebrew-core/blob/master/Formula/bash-completion.rb
==> Caveats
Add the following line to your ~/.bash_profile:
  [[ -r "/usr/local/etc/profile.d/bash_completion.sh" ]] && . "/usr/local/etc/profile.d/bash_completion.sh"

Bash completion has been installed to:
  /usr/local/etc/bash_completion.d
==> Analytics
install: 12,271 (30 days), 39,011 (90 days), 172,085 (365 days)
install-on-request: 11,294 (30 days), 36,102 (90 days), 158,051 (365 days)
build-error: 0 (30 days)
```
Help info:
```shell
Qiannan-MBP:bin qiannan$ pd-ctl completion -h

Output shell completion code for the specified shell (bash).
The shell code must be evaluated to provide interactive
completion of pd-ctl commands.  This can be done by sourcing it from
the .bash_profile.

Usage:
  pd-ctl completion SHELL

Examples:

	# Installing bash completion on macOS using homebrew
	## If running Bash 3.2 included with macOS
	    brew install bash-completion
	## or, if running Bash 4.1+
	    brew install bash-completion@2
	## If you've installed via other means, you may need add the completion to your completion directory
	    pd-ctl completion bash > $(brew --prefix)/etc/bash_completion.d/pd-ctl


	# Installing bash completion on Linux
	## If bash-completion is not installed on Linux, please install the 'bash-completion' package
	## via your distribution's package manager.
	## Load the pd-ctl completion code for bash into the current shell
	    source <(pd-ctl completion bash)
	## Write bash completion code to a file and source if from .bash_profile
		pd-ctl completion bash > ~/completion.bash.inc
		printf "
		# pd-ctl shell completion
		source '$HOME/completion.bash.inc'
		" >> $HOME/.bash_profile
		source $HOME/.bash_profile


Global Flags:
      --cacert string   Path of file that contains list of trusted SSL CAs.
      --cert string     Path of file that contains X509 certificate in PEM format.
  -h, --help            Help message.
      --key string      Path of file that contains X509 key in PEM format.
  -u, --pd string       Address of pd. (default "http://127.0.0.1:2379")
```

Load the completion code to current shell

```shell
source <(pd-ctl completion bash)
```
Type `pd-ctl cl` into the current shell, then type `tab`, it will auto-complete command `pd-ctl cluster`.

Code changes

 - added new file `tools/pd-ctl/pdctl/command/completion_command.go`
 - modified file `tools/pd-ctl/pdctl/ctl.go`

Side effects

 N/A

Related changes
